sdks/java/core/src/main/java/org/apache/beam/sdk/util/PaneInfoTracker.java

```diff
@@ -32,62 +32,109 @@
 
 import org.joda.time.Instant;
 
+import java.util.concurrent.ThreadLocalRandom;
+import javax.annotation.Nullable;
+
 /**
  * Determine the timing and other properties of a new pane for a given computation, key and window.
  * Incorporates any previous pane, whether the pane has been produced because an
  * on-time {@link AfterWatermark} trigger firing, and the relation between the element's timestamp
  * and the current output watermark.
  */
 public class PaneInfoTracker {
+  @VisibleForTesting
+  static final StateTag<Object, ValueState<PaneInfo>> PANE_INFO_TAG =
+      StateTags.makeSystemTagInternal(StateTags.value("pane", PaneInfoCoder.INSTANCE));
+
   private TimerInternals timerInternals;
 
   public PaneInfoTracker(TimerInternals timerInternals) {
     this.timerInternals = timerInternals;
   }
 
-  @VisibleForTesting
-  static final StateTag<Object, ValueState<PaneInfo>> PANE_INFO_TAG =
-      StateTags.makeSystemTagInternal(StateTags.value("pane", PaneInfoCoder.INSTANCE));
+  /**
+   * Could we have any previous pane info stored?
+   */
+  public boolean hasState(ReduceFn<?, ?, ?, ?>.Context context) {
+    return context.window().maxTimestamp().isBefore(GlobalWindow.INSTANCE.maxTimestamp());
```

Comment:
Would love to avoid adding another coupling to `ReduceFn` contexts. It seems the natural parameters are probably the window and the `StateAccessor`, though the latter is not used right now. Any others?